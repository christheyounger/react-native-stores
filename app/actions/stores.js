import {
    GET_STORES,
    GOT_STORES,
    GET_STORE,
    GOT_STORE,
    GET_STORES_FAIL,
    GET_STORE_FAIL,
    GET_STORE_BRANCHES,
    GOT_STORE_BRANCHES,
    GET_STORE_BRANCHES_FAIL
} from '../constants/actiontypes';

export const getStores = () => ({
    type: GET_STORES,
});

export const gotStores = stores => ({
    type: GOT_STORES,
    payload: { stores },
});

export const getStoresFail = error => ({
    type: GET_STORES_FAIL,
    payload: { error },
})

export const getStore = () => ({
    type: GET_STORE,
});

export const gotStore = store => ({
    type: GOT_STORE,
    payload: { store },
});

export const getStoreFail = error => ({
    type: GET_STORE_FAIL,
    payload: { error },
});

export const getStoreBranches = (id) => ({
    type: GET_STORE_BRANCHES,
    payload: { id }
});

export const gotStoreBranches = (branches) => ({
    type: GOT_STORE_BRANCHES,
    payload: { branches },
});

export const getStoreBranchesFail = (error) => ({
    type: GET_STORE_BRANCHES_FAIL,
    payload: { error },
});
