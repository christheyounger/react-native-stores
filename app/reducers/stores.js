import {
    GET_STORES,
    GOT_STORES,
} from '../constants/actiontypes';

const initialState = {
    loading: true,
    stores: []
};

export default function stores(state = initialState, action) {
    switch(action.type) {
        case GET_STORES:
            return initialState;
        case GOT_STORES:
            return {...state, stores: action.payload.stores, loading: false};
        default:
            return state;
    }
}
