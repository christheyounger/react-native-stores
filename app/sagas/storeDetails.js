import { all, put, call, takeLatest } from 'redux-saga/effects'
import settings from "../config/settings";
import { handleErrors } from "../utilities";
import { 
    gotStore,
    getStoreFail,
    getStoreBranches,
    gotStoreBranches,
    getStoreBranchesFail
} from "../actions/stores";
import { GET_STORE, GET_STORE_BRANCHES } from '../constants/actiontypes';

export function* getStore(id) {
    try {
        const response = yield call(fetch, settings.apiUrl + 'store/' + id);
        handleErrors(response);
        const store = response.json();
        yield all([put(gotStore(store)), put(getStoreBranches(store.id))]);
    } catch (error) {
        yield put(getStoreFail(error));
    }
}

export function* watchGetStore() {
    yield takeLatest(GET_STORE, getStore);
}

export function* getStoreBranchesSaga(id) {
    try {
        const response = yield call(fetch, `${settings.apiUrl}'store/${id}/branches`)
        handleErrors(response);
        const branches = response.json();
        yield put(gotStoreBranches(branches));
    } catch (error) {
        yield put(getStoreBranchesFail(error));
    }
}

export function* watchGetBranches() {
    yield takeLatest(GET_STORE_BRANCHES, getStoreBranchesSaga);
}
