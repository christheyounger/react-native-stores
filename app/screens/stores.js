import React from 'react';
import Store from '../components/Store';
import { FlatList, ActivityIndicator, View  } from 'react-native';

export default class StoresList extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount(){
    this.props.getStores();
  }

  handleClick(item) {
    this.props.navigation.navigate('StoreDetails', item);
  }

  render(){
    if(this.props.loading){
      return(
        <View style={{flex: 1, padding: 20}}>
          <ActivityIndicator/>
        </View>
      )
    }
  

    return(
      <View>
        <FlatList
          data={this.props.stores}
          renderItem={({item}) =>
            <Store
              id={item.id}
              name={item.name}
              branches={item.branches}
              handleClick={item => this.handleClick(item)}
            />}
          keyExtractor={({id}) => id.toString()}
        />
      </View>
    );
  }
}
