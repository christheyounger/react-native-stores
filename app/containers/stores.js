import { getStores, gotStores } from "../actions/stores";
import { connect } from 'react-redux';
import StoresList from '../screens/stores';

const matchStateToProps = (state) => {
    const { stores: { loading, stores} } = state;
    return { loading, stores };
};

const mapDispatchToProps = dispatch => ({
    getStores: () => dispatch(getStores()),
    gotStores: payload => dispatch(gotStores(payload)),
});

export default connect(matchStateToProps, mapDispatchToProps)(StoresList);
