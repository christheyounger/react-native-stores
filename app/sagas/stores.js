import { put, call, takeLatest } from 'redux-saga/effects'
import settings from '../config/settings';
import { gotStores, getStoresFail } from '../actions/stores';
import { handleErrors } from '../utilities'

export function* getStores() {
    try {
        const response = yield call(fetch, settings.apiUrl + 'stores')
        handleErrors(response);
        const stores = response.json();
        yield put(gotStores(stores));
    } catch (error) {
        yield put(getStoresFail(error))
    }
}

export function* watchGetStores() {
    yield takeLatest('GET_STORES', getStores)
}