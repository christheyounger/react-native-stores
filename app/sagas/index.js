import { all } from 'redux-saga/effects';
import { watchGetStores } from './stores';
import { watchGetStore, watchGetBranches } from './storeDetails';

export default function* rootSaga() {
    yield all([
        watchGetStores(),
        watchGetStore(),
        watchGetBranches(),
    ])
}
