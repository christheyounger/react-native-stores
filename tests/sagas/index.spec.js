import rootSaga from "../../app/sagas";
import { all } from "redux-saga/effects";
import { watchGetStores } from "../../app/sagas/stores";

describe('rootSaga', () => {
    test('should yield all sagas', () => {
        const saga = rootSaga();    
        expect(saga.next().value).toMatchSnapshot();
    });
})
