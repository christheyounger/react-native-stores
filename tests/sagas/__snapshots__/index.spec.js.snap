// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rootSaga should yield all sagas 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": true,
  "payload": Array [
    GeneratorFunctionPrototype {
      "_invoke": [Function],
    },
  ],
  "type": "ALL",
}
`;
